### PR TITLE
Fix counting of nested named fragments

### PR DESCRIPTION
**Pull request by cotillion**
Fixes graphql-dotnet/graphql-dotnet#3093

It's possible there is a better way to do this but it seems to work. There might be some performance impact..

### CODE REVIEW

**Comment by sungam3r** [comment]
> with the revised code, a recursive query will cause a stack overflow exception.

`NoFragmentCycles` validation rule handles this case. Of course, one can turn it off 😄 .

> As for performance, so long as there are no fragment references, there is no performance impact. And if there is, it is important that the fragments be counted properly.

I agree. One more thing I want to say - it is strange that such an obvious error remained unnoticed. In other places recursion is used when dealing with fragments, and here we just forgot.

> if (count > 500) throw new InvalidOperationException("Too many nested fragments");

I would not. As I said, in other places recursion is also used. Validation rules are intended to catch broken requests. Adding such checks here and there in code base will only complicate and pollute the code.

File: src/GraphQL/Execution/ExecutionStrategy.cs [hunk 1]
```diff
@@ -316,14 +316,27 @@ static void Add(Dictionary<string, (GraphQLField Field, FieldType FieldType)> fi
...
+                    if (selection is GraphQLFragmentSpread spread)
+                    {
                         ++count;
+
+                        var fragment = context.Document.FindFragmentDefinition(spread.FragmentName.Name);
+                        if (fragment != null)
+                        {
+                            count += GetFragmentSpreads(context, fragment.SelectionSet);
+                        }
+
```
Comment (sungam3r) [inline src/GraphQL/Execution/ExecutionStrategy.cs:334]:
```suggestion
```

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
@Shane32 merge if OK.

**Review by Shane32 — APPROVED** [review]

**Comment by Shane32** [comment]
Thanks for the contribution @cotillion !!

**Pull request merged**